products/libllbuild/BuildDB-C-API.cpp

```diff
@@ -0,0 +1,288 @@
+//
+//  BuildDB-C-API.c
+//  libllbuild
+//
+//  Created by Benjamin Herzog on 4/5/19.
+//  Copyright © 2019 Apple Inc. All rights reserved.
+//
+
+#include <llbuild/llbuild.h>
+
+#include "llbuild/Core/BuildDB.h"
+
+#include "llvm/Support/raw_ostream.h"
+
+#include <algorithm>
+#include <cassert>
+#include <memory>
+
+using namespace llbuild;
+using namespace llbuild::core;
+
+namespace {
+
+class CAPIBuildDBResultKeys {
+private:
+  std::unique_ptr<std::map<KeyID, KeyType>> keys;
+  std::unique_ptr<std::map<KeyType, KeyID>> ids;
+  
+public:
+  CAPIBuildDBResultKeys(std::map<KeyID, KeyType> keys): keys(llvm::make_unique<std::map<KeyID, KeyType>>(keys)) {
+    ids = llvm::make_unique<std::map<KeyType, KeyID>>();
+    for (const auto &pair : keys) {
+      (*ids.get())[std::get<1>(pair)] = std::get<0>(pair);
+    }
+  }
+  
+  uint32_t size() {
+    return keys.get()->size();
+  }
+  
+  KeyType keyForID(KeyID index) {
+    return (*keys.get())[index];
+  }
+  
+  KeyID idForKey(KeyType key) {
+    return (*ids.get())[key];
+  }
+};
+  
+class CAPIBuildDBDelegate: public BuildDBDelegate {
+  llb_database_delegate_t cAPIDelegate;
+  
+public:
+  CAPIBuildDBDelegate(llb_database_delegate_t delegate): cAPIDelegate(delegate) { }
+  
+  KeyType getKeyForID(KeyID keyID) override {
```

Comment:
const

Reply:
This overrides an existing definition in BuildDBDelegate which was defined without const, I just replicated that. Should I add it in all other places as well? The BuildSystem for example also implements this.

Reply:
If it is a preexisting problem don't worry about fixing it in this PR